feat: migrate definitions from charm-relation-interfaces

james-garner-canonical opened the pull request:
This PR adds the remaining interface definitions from [charm-relation-interfaces](https://canonical/charm-relation-interfaces) to this repo, in preparation for archiving `charm-relation-interfaces`.

The docs correctly include and display the new interface definitions ([preview build](https://canonical-ubuntu-documentation-library--256.com.readthedocs.build/charmlibs/reference/interfaces/)).

This first commit of this PR adds the files from `charm-relation-interfaces` as-is, as well as copying the existing `tracing` interface's `ruff.toml` file to each interface directory. It may be helpful to [view changes from the subsequent commits only](https://github.com/canonical/charmlibs/pull/256/files/97187c7c6cb1e69ae23230fce7de3e4757cb0110..85c5aa69212ab3b3756f1faee3ff2753d8f64fa5) when reviewing.

Any new interface tests are currently disabled from actually running. However CI will run a job for each interface when merging to main or when changing global files. This is not entirely useless, as it will at least consume the `interface.yaml` files and ensure that they remain valid.

After this PR, the next step should be ensuring that interface library maintainers are aware of the new location, including figuring out what to do with interfaces that aren't used anymore -- there are some interfaces that don't seem to have any charms associated with them on Charmhub.

## Changes

`.scripts/get-interface-test-targets.py` has been updated to skip interface tests with a `.disable` file, and with extra logging.

`.docs/extensions/interface_docs.py`  has been updated to skip including interface docs on the initial passes where we build package reference docs separately.

## New Interfaces

The interfaces are migrated with some minor changes: we run `just format`, adding a `ruff.toml` with some ignores to each interface directory to streamline this, and aligning the directory structure.

I've also taken this opportunity to hyphenate any interface names that are hyphenated according to the charms that use them (they were all normalised to underscores in `charm-relation-interfaces`).

Including the `README.md` files in the docs revealed a few errors with heading levels and malformed `YAML` examples, which I've corrected manually. The readme titles have also been updated to include the interface version.

## Interface Tests

A number of the interface definitions added include interface tests. These have all been disabled for now. However, I also made a few small fixes to the interface test configs: added a branch for one charm, a sub-directory (`charm_root`) for another, and removed a non-existent branch from a third.

james-garner-canonical commented on .docs/extensions/interface_docs.py:
```diff
@@ -37,6 +37,13 @@ def setup(app: sphinx.application.Sphinx) -> dict[str, str | bool]:
+        ref_dir = pathlib.Path(app.confdir, 'reference', 'interfaces')
+        ref_dir.mkdir(parents=True, exist_ok=True)
+        _write_if_needed(path=ref_dir / 'placeholder.md', content='# Temporary TOC placeholder')
+        return
```
This is unfortunately necessary to keep the build time down, as we have `O(mn)` build time, where `m` is the number of packages and `n` is the size of the docs as a whole -- this keeps `n` low during the first `m` passes. It would be great to find a better approach to building the docs eventually.

james-garner-canonical commented on interfaces/azure_storage/interface/v0/interface.yaml:
```diff
@@ -5,6 +5,8 @@ status: draft
   - name: azure-storage-integrator
     url: https://github.com/canonical/object-storage-integrators
+    test_setup:
+      charm_root: azure_storage
```
Manual change.

james-garner-canonical commented on interfaces/etcd_client/interface/v0/interface.yaml:
```diff
@@ -5,6 +5,7 @@ status: draft
 providers:
   - name: charmed-etcd-operator
     url: https://github.com/canonical/charmed-etcd-operator
+    branch: 3.6/edge
```
Manual change.

james-garner-canonical commented on interfaces/profiling/interface/v0/interface.yaml:
```diff
@@ -9,7 +9,6 @@ status: draft
 providers:
   - name: pyroscope-coordinator-k8s
     url: https://github.com/canonical/pyroscope-operators
-    branch: feat/interface-tests-profiling
```
Manual change (branch doesn't exist).

tonyandrewmeyer commented:
> After this PR, the next step should be ensuring that interface library maintainers are aware of the new location, including figuring out what to do with interfaces that aren't used anymore -- there are some interfaces that don't seem to have any charms associated with them on Charmhub.

What's the story with Charmhub and this migration? Will Charmhub still pull from charm-relation-interfaces until that's changed? Is there a plan for us (you) to provide a PR for the store to change that? Some other plan?

tonyandrewmeyer commented on .scripts/get-interface-test-targets.py:
```diff
@@ -94,17 +99,27 @@ def _target_from_interface(
                 continue
+            if not charms:
+                msg = 'No charms for %s %s %s'
+                logger.warning(msg, interface_str, v.name, role)
```
Any reason not to not just do this?


```suggestion
                logger.warning('No charms for %s %s %s', interface_str, v.name, role)
```

tonyandrewmeyer reviewed (commented):
GitHub (or my browser? Or both?) are really struggling with this. Any chance all the automatic formatting could also be moved out to an early commit that could be ignored?

tonyandrewmeyer commented on interfaces/cos_agent/ruff.toml:
```diff
@@ -9,3 +9,10 @@ quote-style = "preserve"
+    "S",  # security
+    "E501",  # line too long
+    "F841",  # assignment to unused variable
+]
```
Ah, I missed it was going into something project-specific rather than being used directly. Ignore then :)

james-garner-canonical commented on .scripts/get-interface-test-targets.py:
```diff
@@ -53,7 +53,12 @@ def _main() -> None:
+    if args.verbose:
+        logger.setLevel(logging.DEBUG)
+    else:
+        logger.setLevel(logging.WARNING)
```
I'll drop the flag from the PR since we can always just edit the verbosity level in the script file itself for debugging -- in CI it runs without the flag (so just warning level).

james-garner-canonical commented:
Thanks. I've dropped the verbosity flag and fixed the `etcd_client` readme title as discussed. I also dropped the `\n` in the mermaid diagram that you pointed out, since it seems to render nice enough (and certainly nicer) without manual line breaks at all.
